### ISSUE TITLE
Model memory leak according to `torch.cuda.memory_allocated` but no matching tracked objects (with example)

### ISSUE THREAD

**Issue by Smerity** [pytorch/pytorch#4890]
This memory issue seems to double the required memory for a model but does so without leaving an obvious trace as to where that memory goes. This is doubly confusing as to double the model size would require doubling the model's weights too. The [example code](https://github.com/Smerity/examples/blob/wlm_memory_bug/word_language_model/main.py#L172) demonstrates this issue using a minimally modified word language modeling example with both LSTM and QRNN (i.e. non-cuDNN based) RNNs.

LSTM:

```
/home/smerity/smer-examples/word_language_model# python main.py --cuda --emsize 2800 --nhid 2800 --nlayers 3 --dropout 0.65 --epochs 40 --tied
Before model run: 1098
Before model run: 2173
Before model run: 2173
```

QRNN (shows issue is independent of cuDNN RNN):

```
/home/smerity/smer-examples/word_language_model# python main.py --cuda --emsize 2800 --nhid 2800 --nlayers 3 --dropout 0.65 --epochs 40 --tied --model QRNN
Before model run: 645
Before model run: 1270
Before model run: 1270
```

When using a helpful [object tree](https://github.com/Smerity/examples/blob/wlm_memory_bug/word_language_model/memtree.py) written by @jekbradbury, we can see (log at bottom) that no extra Python objects are added that refer to tensors or Variables, yet the memory still increases, even when calling `torch.cuda.empty_cache` immediately beforehand.

The example is minimally changed from the Word Language Modeling example and was constructed from a larger model which has the same issue. The changes involve (a) adding QRNN for a non-cuDNN example, (b) moving training of a single batch into its own function so Python objects would be guaranteed to fall out of scope and cause CUDA deletions, and (c) memory diagnostics code.

At this stage I believe something iffy is going on, though at which level I'm not certain, and I also don't rule out my own stupidity :) I also looked at the memory usage before and after every forward/backward via registering a hook and can see these are not buffers that are reused as the memory increases from the "new" baseline, returning to that baseline later.

Happy to help work the issue with anyone who is interested in solving it. Thanks! <3

---

Object tree code (which is commented out by default due to verbosity):

```
/home/smerity/smer-examples/word_language_model# python main.py --cuda --emsize 2800 --nhid 2800 --nlayers 3 --dropout 0.65 --epochs 40 --tied --model QRNN
140349451688920 Parameter of size (8400, 2800) with references held by:
  140348625257064 OrderedDict in which its key is weight
  140348625607560 cell
140348625230328 Parameter of size (8400, 2800) with references held by:
  140348625258696 OrderedDict in which its key is weight
  140348625607560 cell
140348625230568 Parameter of size (8400, 2800) with references held by:
  140348625260328 OrderedDict in which its key is weight
  140348625607560 cell
140348625245256 FloatTensor of size (33278, 2800) with references held by:
  140348625607560 cell
  140349451687640 Parameter of size (33278, 2800)
140349451877832 FloatTensor of size (8400, 2800) with references held by:
  140349451688920 Parameter of size (8400, 2800)
  140348625607560 cell
140348625731848 FloatTensor of size (8400, 2800) with references held by:
  140348625230328 Parameter of size (8400, 2800)
  140348625607560 cell
140348625733064 FloatTensor of size (8400, 2800) with references held by:
  140348625230568 Parameter of size (8400, 2800)
  140348625607560 cell
140349451876168 LongTensor of size (2088628,) with references held by:
  140348625607560 cell
  140349451835504 dict in which its key is train
140349451876552 LongTensor of size (104431, 20) with references held by:
  140348625607560 cell
  140350689706200 dict in which its key is train_data
140349451687640 Parameter of size (33278, 2800) with references held by:
  140348625298840 OrderedDict in which its key is weight
  140348625607560 cell
  140349451881464 OrderedDict in which its key is weight
Before model run: 645
After model run: 1270
Before model run: 2173
140349451688920 Parameter of size (8400, 2800) with references held by:
  140348625257064 OrderedDict in which its key is weight
  140348625607656 cell
140348625230328 Parameter of size (8400, 2800) with references held by:
  140348625258696 OrderedDict in which its key is weight
  140348625607656 cell
140348625230568 Parameter of size (8400, 2800) with references held by:
  140348625260328 OrderedDict in which its key is weight
  140348625607656 cell
140348625245256 FloatTensor of size (33278, 2800) with references held by:
  140348625607656 cell
  140349451687640 Parameter of size (33278, 2800)
140349451877832 FloatTensor of size (8400, 2800) with references held by:
  140349451688920 Parameter of size (8400, 2800)
  140348625607656 cell
140348625731848 FloatTensor of size (8400, 2800) with references held by:
  140348625230328 Parameter of size (8400, 2800)
  140348625607656 cell
140348625733064 FloatTensor of size (8400, 2800) with references held by:
  140348625230568 Parameter of size (8400, 2800)
  140348625607656 cell
140349451876168 LongTensor of size (2088628,) with references held by:
  140348625607656 cell
  140349451835504 dict in which its key is train
140349451876552 LongTensor of size (104431, 20) with references held by:
  140348625607656 cell
  140350689706200 dict in which its key is train_data
140349451687640 Parameter of size (33278, 2800) with references held by:
  140348625298840 OrderedDict in which its key is weight
  140348625607656 cell
  140349451881464 OrderedDict in which its key is weight
Before model run: 1270
After model run: 1270
```

---

The Docker image was built recently (last day or two) from `docker build -t pytorch/pytorch9 -f tools/docker/Dockerfile9 .`

- OS: Linux (Ubuntu)
- PyTorch version: 0.4.0a0+0844b5b
- How you installed PyTorch (conda, pip, source): source
- Python version: 3.6
- CUDA/cuDNN version: release 9.0, V9.0.176
- GPU models and configuration: NVIDIA Volta
- GCC version (if compiling from source): gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.5)

**Comment by apaszke** [comment]
Can you try adding `del output, loss` (in general all intermediates that won't be used in the next iteration of a loop) at the end of [this loop](https://github.com/pytorch/examples/blob/master/word_language_model/main.py#L151)?

I think it's a problem created by Python's scoping rules - `for` loop uses the outer scope and all variables persist across iterations. In this case, the first iteration runs normally, but when you run the second one `output` and `loss` are still keeping the previous iteration graph alive until you actually reassign them.

It shouldn't be a big issue, as we should be dropping all larger buffers progressively during backward, so it might actually be a bug in our current codegen for autograd. Let me know if that change helps!

**Comment by SsnL** [comment]
@apaszke The graph should already be freed at `.backward()`. Could it be an issue of [`repackage_hidden`](https://github.com/Smerity/examples/blob/wlm_memory_bug/word_language_model/main.py#L158)? IIRC, "detach" a variable using `Variable(var.data)` is not encouraged.

**Comment by apaszke** [comment]
Yeah, it shouldn't be doing that, but it can't cause any of the symptoms described here.

**Comment by Smerity** [comment]
Hey @apaszke,

As @SsnL mentioned, I shifted the training of a batch into the separate function [run_train_batch](https://github.com/Smerity/examples/blob/8153be2a2f61d1db2a0e3b7f3968f164ec97a297/word_language_model/main.py#L144) to avoid scoping issues. This ensures all the variables are deleted as they fall out of the `run_train_batch` scope.  The function only returns `hidden` (after having `repackage_hidden(hidden)` called) and a raw float of the loss for that batch.

I experimented with using `del` in my larger model and had the same memory issue.

Just as a sanity check, I inserted this code at the bottom of the `run_train_batch` function:

```
    hidden = repackage_hidden(hidden)
    raw_loss = loss.data
    del output
    del loss
    del data
    del targets
    return hidden, raw_loss
```

and still experienced the same issue:

```
Before model run: 645
After model run: 1270
Before model run: 1270
After model run: 1270
```

For @SsnL, I also tried using `detach` rather than the discouraged method but that had no effect. Admittedly `detach` should be preferred now :)

**Comment by SsnL** [comment]
@Smerity I'll take a look at this tomorrow!

**Comment by SsnL** [comment]
So here is what's going on. After the first backward, every model parameters has a `.grad` attribute. Calling with `zero_grad()` only set its values to zero rather than deleting it. That is why you are seeing exact doubled memory usage after first iteration. You can verify this by changing `model.zero_grad()` to 
```
for p in model.parameters();
  p.grad = None
```
and move it to after `backward()` and before printing.

Whether it should set these `.grad` to `None` is debatable. The current scheme allocates unnecessary memory and launches unnecessary kernels. But `None` and zero conceptually are different. As @ezyang mentioned, it would be great if we can special case zero tensors.

**Comment by Smerity** [comment]
Sorry for the spurious issue, looks like you're quite correct!

I'd assumed that gradients weren't the cause as it wasn't reporting the gradients when we were querying for all Variables / tensors in the [object tree search](https://github.com/Smerity/examples/blob/wlm_memory_bug/word_language_model/memtree.py) and mistakenly assumed `zero_grad` threw the tensors back to the cached memory allocator for use in the next ground.

Thanks for catching what I missed and being so very attentive and rapid in response, @SsnL =]

**Comment by SsnL** [comment]
@Smerity I also assumed that `zero_grad` will free up those memory haha.
